torch/distributed/checkpoint/_fsspec_filesystem.py

```diff
@@ -33,9 +36,29 @@ def create_stream(
         self, path: Union[str, os.PathLike], mode: str
     ) -> Generator[io.IOBase, None, None]:
         assert self.fs is not None
-        with self.fs.transaction:
-            with fsspec.open(str(path), mode) as stream:
+        # _open only supports binary mode
+        if "b" not in mode:
+            with self.create_stream(path, mode.replace("t", "") + "b") as stream:
+                yield io.TextIOWrapper(stream)
+            return
+
+        # fsspec does not support concurrent transactions, so just
+        # manually handle commit/discard for each file.
+        #
+        # This is safe as long as you don't call `create_stream` on
+        # the same path concurrently
+        autocommit = "r" in mode
```

Comment:
Unfortunately, running `stream.commit()` doesn't always work when things are set to read mode: for instance ,with a local file system, https://github.com/fsspec/filesystem_spec/blob/76ca4a68885d572880ac6800f079738df562f02c/fsspec/implementations/local.py#L368 special-cases the code such that you can't actually `commit` on read mode (since `self.temp` will not be declared).

I cold modify this to check that `mode != "rb"` I guess to be more exact? I agree that `r+` also exists...

Reply:
There is indeed a mismatch here, due to most remote filesystems only supporting "w", "r" and maybe "a". Files should have a `.writable()` method, but of course that's only present after open.

Reply:
I think `.writable()` is exactly what we need here actually -- I can just set `autocommit=False` unconditionally and only run `.commit` or `.discard` if the stream is writable.